File: lib/core/hiring/work_schedule_time.dart
/// 공고·합의에 포함된 근무 시각 파싱 (MVP)
abstract final class WorkScheduleTime {
  static const checkGraceAfterEnd = Duration(hours: 2);

  /// `09:00–18:00`, `09:00-18:00`, `09:00~18:00` 등에서 종료 시각
  static (int hour, int minute)? parseEndClock(String workSchedule) {
    final match = RegExp(r'(\d{1,2}):(\d{2})\s*$').firstMatch(workSchedule.trim());
    if (match == null) {
      final range = RegExp(
        r'(\d{1,2}):(\d{2})\s*[~\-–]\s*(\d{1,2}):(\d{2})',
      ).firstMatch(workSchedule);
      if (range == null) return null;
      return (int.parse(range.group(3)!), int.parse(range.group(4)!));
    }
    return (int.parse(match.group(1)!), int.parse(match.group(2)!));
  }

  static (int hour, int minute)? parseStartClock(String workSchedule) {
    final range = RegExp(
      r'(\d{1,2}):(\d{2})\s*[~\-–]\s*(\d{1,2}):(\d{2})',
    ).firstMatch(workSchedule);
    if (range == null) return null;
    return (int.parse(range.group(1)!), int.parse(range.group(2)!));
  }

  static DateTime? workStartAt(DateTime workDate, String workSchedule) {
    final clock = parseStartClock(workSchedule);
    if (clock == null) return null;
    return DateTime(workDate.year, workDate.month, workDate.day, clock.$1, clock.$2);
  }

  static DateTime? workEndAt(DateTime workDate, String workSchedule) {
    final clock = parseEndClock(workSchedule);
    if (clock == null) return null;
    return DateTime(workDate.year, workDate.month, workDate.day, clock.$1, clock.$2);
  }

  static DateTime? checkWindowClosesAt(DateTime workDate, String workSchedule) {
    final end = workEndAt(workDate, workSchedule);
    return end?.add(checkGraceAfterEnd);
  }

  /// 출근까지 남은 시간 — `1일 9h 45m`
  static String? countdownLabel({
    required DateTime? workDate,
    required String workSchedule,
    DateTime? now,
  }) {
    if (workDate == null) return null;
    final start = workStartAt(workDate, workSchedule);
    if (start == null) return null;
    final current = now ?? DateTime.now();
    if (!current.isBefore(start)) return null;
    return _formatDuration(start.difference(current));
  }

  static String _formatDuration(Duration d) {
    final days = d.inDays;
    final hours = d.inHours.remainder(24);
    final minutes = d.inMinutes.remainder(60);
    if (days > 0) return '${days}일 ${hours}h ${minutes}m';
    if (hours > 0) return '${hours}h ${minutes}m';
    return '${minutes}m';
  }
}

File: server/app/services/work_schedule_time.py
"""공고·합의에 포함된 근무 시각 파싱 — lib/core/hiring/work_schedule_time.dart 포팅."""

from __future__ import annotations

import re
from datetime import date, datetime

_RANGE_RE = re.compile(r"(\d{1,2}):(\d{2})\s*[~\-–]\s*(\d{1,2}):(\d{2})")
_CLOCK_RE = re.compile(r"(\d{1,2}):(\d{2})")


def parse_start_clock(work_schedule: str) -> tuple[int, int] | None:
    """`09:00–18:00`, `09:00-18:00`, `09:00~18:00` 등에서 시작 시각."""
    match = _RANGE_RE.search(work_schedule or "")
    if match is None:
        return None
    return int(match.group(1)), int(match.group(2))


def parse_clock(value: str) -> tuple[int, int] | None:
    """`09:30` 같은 단일 시각 문자열 파싱 — 셔틀 탑승 시각 등."""
    match = _CLOCK_RE.search(value or "")
    if match is None:
        return None
    return int(match.group(1)), int(match.group(2))


def work_start_at(work_date: date, work_schedule: str) -> datetime | None:
    clock = parse_start_clock(work_schedule)
    if clock is None:
        return None
    hour, minute = clock
    return datetime(work_date.year, work_date.month, work_date.day, hour, minute)


def combine_date_and_clock(the_date: date, clock_text: str) -> datetime | None:
    clock = parse_clock(clock_text)
    if clock is None:
        return None
    hour, minute = clock
    return datetime(the_date.year, the_date.month, the_date.day, hour, minute)
